charts/newrelic-infrastructure-v3/templates/_helpers_compatibility.tpl

```diff
@@ -0,0 +1,79 @@
+{{/*
+Returns true if .Values.ksm.enabled is true and the legacy disableKubeStateMetrics is not set
+*/}}
+{{- define "newrelic.compatibility.ksm.enabled" -}}
+{{- if and .Values.ksm.enabled (not .Values.disableKubeStateMetrics) -}}
+true
+{{- end -}}
+{{- end -}}
+
+{{/*
+Returns legacy ksm values
+*/}}
+{{- define "newrelic.compatibility.ksm.legacyData" -}}
+enabled: true
+{{- if .Values.kubeStateMetricsScheme }}
+scheme: {{ .Values.kubeStateMetricsScheme }}
+{{- end -}}
+{{- if .Values.kubeStateMetricsPort }}
+port: {{ .Values.kubeStateMetricsPort }}
+{{- end -}}
+{{- if .Values.kubeStateMetricsUrl }}
+staticURL: {{ .Values.kubeStateMetricsUrl }}
+{{- end -}}
+{{- if .Values.kubeStateMetricsPodLabel }}
+selector: {{ printf "%s=kube-state-metrics" .Values.kubeStateMetricsPodLabel }}
+{{- end -}}
+{{- if  .Values.kubeStateMetricsNamespace }}
+namespace: {{ .Values.kubeStateMetricsNamespace}}
+{{- end -}}
+{{- end -}}
+
+{{/*
+Returns the new value if available, falling back on the legacy one
+*/}}
+{{- define "newrelic.compatibility.valueWithFallback" -}}
+{{- if .supported }}
+{{- toYaml .supported}}
+{{- else if .legacy -}}
+{{- toYaml .legacy}}
+{{- end }}
+{{- end -}}
+
+{{/*
+Returns securityContext merged with old runAsUser config
+*/}}
+{{- define "newrelic.compatibility.securityContext" -}}
+{{- if  .Values.runAsUser -}}
+{{- mustMergeOverwrite .Values.securityContext (dict "runAsUser" .Values.runAsUser ) | toYaml }}
```

Comment:
We are already calling `mustMergeOverwrite` outside of the helper, do we need to call it inside as well?

Reply:
We cannot write directly:
```
{{- mustMergeOverwrite .Values.securityContext (dict "runAsUser" .Values.runAsUser ) | toYaml | nindent 12 }}
```
since it would result in a null in case runAsUser is not set, therefore we would need an if merging it in one case and ot in the other, making it really verbose.

```
                        Expected to equal:
                                1000
                        Actual:
                                null
                        Diff:
                                --- Expected
                                +++ Actual
                                @@ -1,2 +1,2 @@
                                -1000
                                +null

```

We could remove easily the mustMergeOverwrite outside the helper, but we would lose the concept of overlay

Reply:
not at all, I was following the idea that the overlay needed to be a valid config